packages/web3-shared/src/hooks/useERC20TokenDetailed.ts

```diff
@@ -1,66 +1,115 @@
-        [erc20TokenContract, chainId],
-    )
+    return {
+        loading: asyncToken.loading,
+        error: asyncToken.error ?? null,
+        value: tokenDetailed,
+    } as AsyncStateRetry<ERC20TokenDetailed>
+}
 
-    // token detailed
-    const tokenDetailed = useMemo(() => {
-        return {
-            type: EthereumTokenType.ERC20,
-            address: formatEthereumAddress(address),
-            chainId,
-            symbol: parseStringOrBytes32(tokenSymbol, tokenSymbolBytes32, 'UNKNOWN'),
-            name: parseStringOrBytes32(tokenName, tokenNameBytes32, 'Unknown Token'),
-            decimals: typeof tokenDecimals === 'string' ? Number.parseInt(tokenDecimals, 10) : tokenDecimals,
-        } as ERC20TokenDetailed
-    }, [tokenName, tokenNameBytes32, tokenSymbol, tokenSymbolBytes32, tokenDecimals, chainId])
+export function useERC20TokensDetailed(listOfToken: Pick<ERC20Token, 'type' | 'address'>[]) {
+    const chainId = useChainId()
+    const listOfAddress = useMemo(() => listOfToken.map((t) => t.address), [JSON.stringify(listOfToken)])
+    const erc20TokenContracts = useERC20TokenContracts(listOfAddress)
+    const erc20TokenBytes32Contracts = useERC20TokenBytes32Contracts(listOfAddress)
 
-    const asyncList = [
-        asyncTokenName,
-        asyncTokenNameBytes32,
-        asyncTokenSymbol,
-        asyncTokenSymbolBytes32,
-        asyncTokenDecimals,
-    ]
+    const { value: tokens, ...asyncTokens } = useAsyncRetry(async () => {
+        return Promise.allSettled(
+            listOfToken.map(async (token, i) => {
+                const erc20TokenContract = erc20TokenContracts[i]
+                const erc20TokenBytes32Contract = erc20TokenBytes32Contracts[i]
+                const _token = await getToken(erc20TokenContract, erc20TokenBytes32Contract, token)
+                return { address: token.address, ..._token }
+            }),
+        )
+    }, [
+        chainId,
+        JSON.stringify(erc20TokenContracts),
+        JSON.stringify(erc20TokenBytes32Contracts),
+        JSON.stringify(listOfToken),
+    ])
+
+    const tokensDetailed = useMemo(
+        () =>
+            tokens
+                ? tokens.map((token, i) =>
+                      token.status === 'fulfilled'
+                          ? createERC20Token(
+                                chainId,
+                                token.value.address,
+                                typeof token.value.decimals === 'string'
+                                    ? Number(token.value.decimals)
+                                    : token.value.decimals,
+                                parseStringOrBytes32(token.value.name, token.value.nameBytes32, 'Unknown Token'),
+                                parseStringOrBytes32(token.value.symbol, token.value.symbolBytes32, 'Unknown'),
+                            )
+                          : createLoadingFailToken(listOfToken[i].address, chainId, token.reason),
```

Comment:
Why do we need loading to fail token here? It's not a token at all.

Reply:
Save the address of fail token to retry it again.